tests/php/Dev/DeprecationTest.php

```diff
@@ -2,16 +2,40 @@
 
 namespace SilverStripe\Dev\Tests;
 
+use PHPUnit\Framework\Error\Deprecated;
+use SilverStripe\Core\Config\Config;
 use SilverStripe\Dev\Deprecation;
 use SilverStripe\Dev\SapphireTest;
-use SilverStripe\Dev\Tests\DeprecationTest\TestDeprecation;
+use SilverStripe\Dev\Tests\DeprecationTest\DeprecationTestObject;
 
 class DeprecationTest extends SapphireTest
 {
+    protected static $extra_dataobjects = [
+        DeprecationTestObject::class,
+    ];
+
+    private $oldHandler;
+
+    protected function setup(): void
+    {
+        // Use custom error handler for two reasons:
+        // - Filter out errors for deprecated class properities unrelated to this unit test
+        // - Allow the use of expectDeprecation(), which doesn't work with E_USER_DEPRECATION by default
+        //   https://github.com/laminas/laminas-di/pull/30#issuecomment-927585210
+        parent::setup();
+        $this->oldHandler = set_error_handler(function ($errno, $errstr) {
+            if (str_contains($errstr, 'SilverStripe\\Dev\\Tests\\DeprecationTest')) {
+                throw new Deprecated($errstr, $errno, '', 1);
+            } else {
+                // noop
+            }
+        }, E_USER_DEPRECATED);
+    }
+
     protected function tearDown(): void
     {
-        Deprecation::$notice_level = E_USER_DEPRECATED;
         Deprecation::disable();
+        set_error_handler($this->oldHandler);
```

Comment:
```suggestion
        restore_error_handler();
```
We don't want to add it onto the handler stack - we just want to pop our handler _off_ the stack.

Reply:
I'm not sure what a "handler stack" is? Can you add multiple error handlers deep, and thus restore multiple errors handlers deep? I don't see any reference to that in https://www.php.net/manual/en/function.set-error-handler.php?

If we used `restore_error_handler();` here, what would happen if there is another call to `set_error_handler()` inside code called by one of the unit tests, and an Exception/Error is thrown, and that error handler is not restored.  In that scenario `restore_error_handler()` would restore to the "inner" handler, and not `$this->oldHandler`

Reply:
The way I understand it is that there is a stack, and when you call `set_error_handler()` you add a handler to the stack. When you call `restore_error_handler()` you pop your handler off the stack.

This means that if you call `set_error_handler()`, then call some third party code which calls `set_error_handler()` and then `restore_error_handler()`, you can be assured that calling `restore_error_handler()` after your code will result in the handler that was active _before_ your code being active again.

Regardless of how the internals actually do this (i.e. with a stack or some other data structure), this definitely appears to be the way it works in practice.

> If we used `restore_error_handler();` here, what would happen if there is another call to `set_error_handler()` inside code called by one of the unit tests, and an Exception/Error is thrown, and that error handler is not restored

`restore_error_handler()` restores the previous handler. So in that scenario the current handler after this call would be the one added in setup. But in that case the test is setup poorly. Is that even an actual scenario though? Do we have any tests that call `set_error_handler()`? If so that test should probably be updated to use a try finally pattern, or use `set_error_handler()` in setup and `restore_error_handler()` in teardown.

Reply:
> Do we have any tests that call set_error_handler()

It's not tests that call `set_error_handler()`, it's code that's within tests calling `set_error_handler()`,.  I'm mostly sure that code in `Config::get()` currently will not set a custom error handler, however in the future I cannot give an guarantees.

> If so that test should probably be updated to use a try finally pattern, or use set_error_handler() in setup and restore_error_handler() in teardown.

Well, these tests ARE currently are using setup() + teardown() :-)

Reply:
I think my recommendation is the correct way to do this - but I also don't think it's worth holding back this PR because of this as ultimately in the real world neither implementation is likely to cause us any trouble. I'll just mark this as resolved.